File: utils/segmentDownloader.js
#!/usr/bin/env node
/**
 * Segmented Audio Downloader for YouTube Videos (Node.js Version)
 * ===============================================================
 * 
 * This utility downloads audio segments from a segmented streaming backend
 * and optionally combines them into a single file using ffmpeg.
 * 
 * Usage:
 *     node segmentDownloader.js <video_id> [options]
 * 
 * Features:
 * - Downloads segments progressively (first 15s, then 30s chunks)
 * - Monitors backend /status endpoint for segment availability  
 * - Robust error handling with retries
 * - Optional ffmpeg concatenation
 * - Progress tracking and logging
 * 
 * Requirements:
 *     npm install axios cli-progress commander
 *     # Optional: ffmpeg for concatenation
 */

const fs = require('fs').promises;
const path = require('path');
const { pipeline } = require('stream').promises;
const { createWriteStream } = require('fs');
const { spawn } = require('child_process');
const axios = require('axios');
const { Command } = require('commander');

// Optional progress bar
let ProgressBar;
try {
  const cliProgress = require('cli-progress');
  ProgressBar = cliProgress.SingleBar;
} catch (e) {
  console.log('Install cli-progress for progress bars: npm install cli-progress');
}

// Configuration
const config = {
  BACKEND_URL: 'https://functions-offers-audit-insertion.trycloudflare.com',
  DOWNLOAD_DIR: './downloads',
  MAX_CONCURRENT: 3,
  POLLING_INTERVAL: 2000, // milliseconds
  MAX_RETRIES: 3,
  REQUEST_TIMEOUT: 30000,
  USER_AGENT: 'Mozilla/5.0 (compatible; SegmentDownloader/1.0)'
};

// Logging utility
class Logger {
  constructor(verbose = false) {
    this.verbose = verbose;
  }

  log(level, message, ...args) {
    const timestamp = new Date().toISOString();
    const fullMessage = `${timestamp} - ${level.toUpperCase()} - ${message}`;
    
    if (level === 'error' || level === 'warn' || level === 'info' || this.verbose) {
      console.log(fullMessage, ...args);
    }
  }

  info(message, ...args) { this.log('info', message, ...args); }
  debug(message, ...args) { this.log('debug', message, ...args); }
  warn(message, ...args) { this.log('warn', message, ...args); }
  error(message, ...args) { this.log('error', message, ...args); }
}

const logger = new Logger();

/**
 * Main class for handling segmented audio downloads
 */
class SegmentDownloader {
  constructor(videoId, backendUrl = config.BACKEND_URL) {
    this.videoId = videoId;
    this.backendUrl = backendUrl;
    this.downloadDir = path.join(config.DOWNLOAD_DIR, videoId);
    
    // Track download state
    this.downloadedSegments = new Set();
    this.failedSegments = new Set();
    
    // Configure axios
    this.client = axios.create({
      timeout: config.REQUEST_TIMEOUT,
      headers: { 'User-Agent': config.USER_AGENT }
    });

    logger.info(`Initialized downloader for video: ${videoId}`);
    logger.info(`Backend URL: ${this.backendUrl}`);
    logger.info(`Download directory: ${this.downloadDir}`);
  }

  /**
   * Initialize download directory
   */
  async init() {
    try {
      await fs.mkdir(this.downloadDir, { recursive: true });
    } catch (error) {
      logger.error(`Failed to create download directory: ${error.message}`);
      throw error;
    }
  }

  /**
   * Check which segments are available on the backend
   * @returns {Promise<Object>} Status object with ready segments
   */
  async getSegmentStatus() {
    try {
      const url = `${this.backendUrl}/status?v=${this.videoId}`;
      const response = await this.client.get(url);
      
      logger.debug(`Status response:`, response.data);
      
      return {
        ready: response.data.ready || [],
        total: response.data.totalSegments
      };
      
    } catch (error) {
      logger.error(`Failed to get segment status: ${error.message}`);
      return { ready: [], total: null };
    }
  }

  /**
   * Download a specific segment
   * @param {number} segmentNum - Segment number to download
   * @param {number} retryCount - Current retry attempt
   * @returns {Promise<boolean>} True if successful
   */
  async downloadSegment(segmentNum, retryCount = 0) {
    const segmentFile = path.join(this.downloadDir, `segment_${segmentNum}.m4a`);
    
    // Skip if already exists and is valid
    try {
      const stats = await fs.stat(segmentFile);
      if (stats.size > 0) {
        logger.info(`Segment ${segmentNum} already exists, skipping`);
        this.downloadedSegments.add(segmentNum);
        return true;
      }
    } catch (error) {
      // File doesn't exist, continue with download
    }

    try {
      const url = `${this.backendUrl}/stream?v=${this.videoId}&segment=${segmentNum}`;
      logger.info(`Downloading segment ${segmentNum} from ${url}`);
      
      const response = await this.client.get(url, { 
        responseType: 'stream',
        timeout: config.REQUEST_TIMEOUT 
      });
      
      // Check content type
      const contentType = response.headers['content-type'] || '';
      if (!['audio', 'video', 'octet-stream'].some(ct => contentType.includes(ct))) {
        logger.warn(`Unexpected content type for segment ${segmentNum}: ${contentType}`);
      }

      // Setup progress bar if available
      let progressBar;
      const contentLength = parseInt(response.headers['content-length'] || '0');
      if (ProgressBar && contentLength > 0) {
        progressBar = new ProgressBar({
          format: `Segment ${segmentNum} |{bar}| {percentage}% | {value}/{total} bytes`,
          barCompleteChar: '\u2588',
          barIncompleteChar: '\u2591',
          hideCursor: true
        });
        progressBar.start(contentLength, 0);
      }

      // Download with progress tracking
      const writer = createWriteStream(segmentFile);
      let downloadedBytes = 0;

      response.data.on('data', (chunk) => {
        downloadedBytes += chunk.length;
        if (progressBar) {
          progressBar.update(downloadedBytes);
        }
      });

      // Use pipeline for proper error handling
      await pipeline(response.data, writer);

      if (progressBar) {
        progressBar.stop();
      }

      // Verify download
      const stats = await fs.stat(segmentFile);
      if (stats.size === 0) {
        await fs.unlink(segmentFile);
        throw new Error(`Downloaded segment ${segmentNum} is empty`);
      }

      logger.info(`Successfully downloaded segment ${segmentNum} (${stats.size} bytes)`);
      this.downloadedSegments.add(segmentNum);
      return true;

    } catch (error) {
      logger.error(`Error downloading segment ${segmentNum}: ${error.message}`);
      
      // Handle retries
      if (retryCount < config.MAX_RETRIES) {
        const waitTime = Math.pow(2, retryCount) * 1000; // Exponential backoff
        logger.info(`Retrying segment ${segmentNum} in ${waitTime/1000} seconds (attempt ${retryCount + 1}/${config.MAX_RETRIES})`);
        
        await new Promise(resolve => setTimeout(resolve, waitTime));
        return this.downloadSegment(segmentNum, retryCount + 1);
      } else {
        logger.error(`Failed to download segment ${segmentNum} after ${config.MAX_RETRIES} attempts`);
        this.failedSegments.add(segmentNum);
        return false;
      }
    }
  }

  /**
   * Download all currently available segments
   * @param {number?} maxSegments - Maximum number of segments to download
   * @returns {Promise<number[]>} Array of successfully downloaded segment numbers
   */
  async downloadAvailableSegments(maxSegments = null) {
    const status = await this.getSegmentStatus();
    const availableSegments = status.ready;

    if (availableSegments.length === 0) {
      logger.warn('No segments available for download');
      return [];
    }

    // Filter out already downloaded/failed segments
    let toDownload = availableSegments.filter(seg => 
      !this.downloadedSegments.has(seg) && !this.failedSegments.has(seg)
    );

    if (maxSegments) {
      toDownload = toDownload.slice(0, maxSegments);
    }

    if (toDownload.length === 0) {
      logger.info('All available segments already downloaded');
      return Array.from(this.downloadedSegments).sort((a, b) => a - b);
    }

    logger.info(`Downloading ${toDownload.length} segments: [${toDownload.join(', ')}]`);

    // Download segments with concurrency limit
    const successfulDownloads = [];
    const downloadPromises = [];
    
    for (let i = 0; i < toDownload.length; i += config.MAX_CONCURRENT) {
      const batch = toDownload.slice(i, i + config.MAX_CONCURRENT);
      
      const batchPromises = batch.map(async (segmentNum) => {
        const success = await this.downloadSegment(segmentNum);
        if (success) {
          successfulDownloads.push(segmentNum);
        }
        return success;
      });

      await Promise.all(batchPromises);
    }

    return successfulDownloads.sort((a, b) => a - b);
  }

  /**
   * Continuously monitor for new segments and download them
   * @param {number} maxWaitTime - Maximum time to wait in milliseconds
   * @returns {Promise<number[]>} Array of all downloaded segment numbers
   */
  async monitorAndDownload(maxWaitTime = 300000) {
    logger.info(`Starting continuous monitoring (max wait: ${maxWaitTime/1000}s)`);
    
    const startTime = Date.now();
    let lastSegmentCount = 0;
    let noNewSegmentsCount = 0;

    while (Date.now() - startTime < maxWaitTime) {
      try {
        // Download currently available segments
        const downloaded = await this.downloadAvailableSegments();

        if (downloaded.length > lastSegmentCount) {
          lastSegmentCount = downloaded.length;
          noNewSegmentsCount = 0;
          logger.info(`Downloaded ${downloaded.length} segments so far`);
        } else {
          noNewSegmentsCount++;
        }

        // Check if we should continue monitoring
        const status = await this.getSegmentStatus();
        if (status.total && this.downloadedSegments.size >= status.total) {
          logger.info('All segments downloaded according to backend');
          break;
        }

        // Stop if no new segments for too long
        if (noNewSegmentsCount >= 10) { // 20 seconds without new segments
          logger.info('No new segments detected for extended period, stopping monitor');
          break;
        }

        // Wait before next check
        logger.debug(`Waiting ${config.POLLING_INTERVAL/1000}s before next check...`);
        await new Promise(resolve => setTimeout(resolve, config.POLLING_INTERVAL));

      } catch (error) {
        logger.error(`Error during monitoring: ${error.message}`);
        await new Promise(resolve => setTimeout(resolve, config.POLLING_INTERVAL));
      }
    }

    logger.info(`Monitoring completed. Downloaded ${this.downloadedSegments.size} segments`);
    return Array.from(this.downloadedSegments).sort((a, b) => a - b);
  }

  /**
   * Combine downloaded segments into a single audio file
   * @param {string} outputFile - Path to output file
   * @param {boolean} useFFmpeg - Whether to use ffmpeg
   * @returns {Promise<boolean>} True if successful
   */
  async combineSegments(outputFile, useFFmpeg = true) {
    if (this.downloadedSegments.size === 0) {
      logger.error('No segments downloaded to combine');
      return false;
    }

    const segmentFiles = [];
    for (const segNum of Array.from(this.downloadedSegments).sort((a, b) => a - b)) {
      const segmentFile = path.join(this.downloadDir, `segment_${segNum}.m4a`);
      try {
        await fs.access(segmentFile);
        segmentFiles.push(segmentFile);
      } catch (error) {
        logger.warn(`Segment file not found: ${segmentFile}`);
      }
    }

    if (segmentFiles.length === 0) {
      logger.error('No segment files found for combination');
      return false;
    }

    logger.info(`Combining ${segmentFiles.length} segments into ${outputFile}`);

    if (useFFmpeg && await this._hasFFmpeg()) {
      return this._combineWithFFmpeg(segmentFiles, outputFile);
    } else {
      return this._combineSimple(segmentFiles, outputFile);
    }
  }

  /**
   * Check if ffmpeg is available
   */
  async _hasFFmpeg() {
    try {
      await new Promise((resolve, reject) => {
        const child = spawn('ffmpeg', ['-version']);
        child.on('close', (code) => {
          if (code === 0) resolve();
          else reject(new Error('ffmpeg not found'));
        });
        child.on('error', reject);
      });
      return true;
    } catch (error) {
      return false;
    }
  }

  /**
   * Combine segments using ffmpeg
   */
  async _combineWithFFmpeg(segmentFiles, outputFile) {
    try {
      // Create file list for ffmpeg
      const listFile = path.join(this.downloadDir, 'segments_list.txt');
      const listContent = segmentFiles
        .map(file => `file '${path.resolve(file)}'`)
        .join('\n');
      
      await fs.writeFile(listFile, listContent);

      // Run ffmpeg
      const args = [
        '-y', '-f', 'concat', '-safe', '0',
        '-i', listFile, '-c', 'copy', outputFile
      ];

      logger.info(`Running ffmpeg: ffmpeg ${args.join(' ')}`);

      await new Promise((resolve, reject) => {
        const child = spawn('ffmpeg', args);
        
        let stderr = '';
        child.stderr.on('data', (data) => {
          stderr += data.toString();
        });

        child.on('close', (code) => {
          if (code === 0) {
            resolve();
          } else {
            reject(new Error(`FFmpeg failed with code ${code}: ${stderr}`));
          }
        });

        child.on('error', reject);
      });

      // Clean up
      await fs.unlink(listFile);

      logger.info(`Successfully combined segments with ffmpeg: ${outputFile}`);
      return true;

    } catch (error) {
      logger.error(`FFmpeg combination failed: ${error.message}`);
      return false;
    }
  }

  /**
   * Simple binary concatenation (may not work for all formats)
   */
  async _combineSimple(segmentFiles, outputFile) {
    try {
      const outputHandle = await fs.open(outputFile, 'w');

      for (const segmentFile of segmentFiles) {
        const data = await fs.readFile(segmentFile);
        await outputHandle.write(data);
      }

      await outputHandle.close();

      logger.info(`Successfully combined segments (simple): ${outputFile}`);
      return true;

    } catch (error) {
      logger.error(`Simple combination failed: ${error.message}`);
      return false;
    }
  }

  /**
   * Get download statistics
   */
  getDownloadStats() {
    return {
      videoId: this.videoId,
      downloadedSegments: this.downloadedSegments.size,
      failedSegments: this.failedSegments.size,
      downloadedList: Array.from(this.downloadedSegments).sort((a, b) => a - b),
      failedList: Array.from(this.failedSegments).sort((a, b) => a - b),
      downloadDirectory: this.downloadDir
    };
  }
}

/**
 * Main CLI function
 */
async function main() {
  const program = new Command();
  
  program
    .name('segment-downloader')
    .description('Download segmented audio from YouTube video')
    .version('1.0.0')
    .argument('<video-id>', 'YouTube video ID (11 characters)')
    .option('--backend-url <url>', 'Backend server URL', config.BACKEND_URL)
    .option('--download-dir <dir>', 'Download directory', config.DOWNLOAD_DIR)
    .option('--max-wait <seconds>', 'Max wait time in seconds', '300')
    .option('--no-monitor', 'Download available segments only, no monitoring')
    .option('--combine <file>', 'Combine segments into output file')
    .option('--no-ffmpeg', 'Use simple concatenation instead of ffmpeg')
    .option('--concurrent <num>', 'Max concurrent downloads', config.MAX_CONCURRENT.toString())
    .option('--verbose', 'Verbose logging')
    .parse();

  const options = program.opts();
  const videoId = program.args[0];

  // Configure logging
  if (options.verbose) {
    logger.verbose = true;
  }

  // Validate video ID
  if (!videoId || videoId.length !== 11) {
    logger.error('Invalid video ID. Must be 11 characters (YouTube video ID)');
    process.exit(1);
  }

  // Update configuration
  config.BACKEND_URL = options.backendUrl;
  config.DOWNLOAD_DIR = options.downloadDir;
  config.MAX_CONCURRENT = parseInt(options.concurrent);

  try {
    // Create and initialize downloader
    const downloader = new SegmentDownloader(videoId, options.backendUrl);
    await downloader.init();

    let downloaded;
    if (options.noMonitor) {
      // Download available segments only
      logger.info('Downloading currently available segments...');
      downloaded = await downloader.downloadAvailableSegments();
    } else {
      // Monitor and download continuously
      logger.info('Starting continuous monitoring and download...');
      const maxWaitMs = parseInt(options.maxWait) * 1000;
      downloaded = await downloader.monitorAndDownload(maxWaitMs);
    }

    // Show statistics
    const stats = downloader.getDownloadStats();
    logger.info(`Download complete:`, stats);

    // Combine segments if requested
    if (options.combine && downloaded.length > 0) {
      const success = await downloader.combineSegments(options.combine, !options.noFfmpeg);
      if (success) {
        logger.info(`Audio combined successfully: ${options.combine}`);
      } else {
        logger.error('Failed to combine segments');
        process.exit(1);
      }
    }

    if (downloaded.length === 0) {
      logger.warn('No segments were downloaded');
      process.exit(1);
    }

  } catch (error) {
    logger.error(`Unexpected error: ${error.message}`);
    process.exit(1);
  }
}

// Handle process signals
process.on('SIGINT', () => {
  logger.info('Download interrupted by user');
  process.exit(1);
});

process.on('SIGTERM', () => {
  logger.info('Download terminated');
  process.exit(1);
});

// Export for use as module
module.exports = { SegmentDownloader, Logger, config };

// Run as CLI if this file is executed directly
if (require.main === module) {
  main().catch(error => {
    console.error('Fatal error:', error);
    process.exit(1);
  });
}

/*
Example usage:

# Basic download with monitoring
node segmentDownloader.js dQw4w9WgXcQ --verbose

# Download and combine with ffmpeg
node segmentDownloader.js dQw4w9WgXcQ --combine output.m4a --verbose

# Download available segments only (no monitoring)
node segmentDownloader.js dQw4w9WgXcQ --no-monitor --concurrent 5

# Custom backend URL and download directory  
node segmentDownloader.js dQw4w9WgXcQ --backend-url https://custom-backend.com --download-dir ./music

# Simple concatenation without ffmpeg
node segmentDownloader.js dQw4w9WgXcQ --combine output.m4a --no-ffmpeg
*/

File: utils/segmentDownloader.py
#!/usr/bin/env python3
"""
Segmented Audio Downloader for YouTube Videos
==============================================

This utility downloads audio segments from a segmented streaming backend
and optionally combines them into a single file using ffmpeg.

Usage:
    python segmentDownloader.py <video_id> [options]

Features:
- Downloads segments progressively (first 15s, then 30s chunks)
- Monitors backend /status endpoint for segment availability
- Robust error handling with retries
- Optional ffmpeg concatenation
- Progress tracking and logging

Requirements:
    pip install requests tqdm
    # Optional: ffmpeg for concatenation
"""

import os
import sys
import time
import json
import argparse
import logging
from pathlib import Path
from typing import List, Dict, Optional, Set
import requests
from concurrent.futures import ThreadPoolExecutor, as_completed
from dataclasses import dataclass

try:
    from tqdm import tqdm
    HAS_TQDM = True
except ImportError:
    HAS_TQDM = False
    print("Install tqdm for progress bars: pip install tqdm")

# Configuration
@dataclass
class Config:
    BACKEND_URL: str = "https://functions-offers-audit-insertion.trycloudflare.com"
    DOWNLOAD_DIR: str = "./downloads"
    MAX_CONCURRENT: int = 3
    POLLING_INTERVAL: float = 2.0  # seconds
    MAX_RETRIES: int = 3
    REQUEST_TIMEOUT: int = 30
    USER_AGENT: str = "Mozilla/5.0 (compatible; SegmentDownloader/1.0)"

# Global config instance
config = Config()

# Setup logging
logging.basicConfig(
    level=logging.INFO,
    format='%(asctime)s - %(levelname)s - %(message)s',
    handlers=[
        logging.FileHandler('segment_downloader.log'),
        logging.StreamHandler(sys.stdout)
    ]
)
logger = logging.getLogger(__name__)

class SegmentDownloadError(Exception):
    """Custom exception for segment download errors"""
    pass

class SegmentDownloader:
    """Main class for handling segmented audio downloads"""
    
    def __init__(self, video_id: str, backend_url: str = None):
        self.video_id = video_id
        self.backend_url = backend_url or config.BACKEND_URL
        self.download_dir = Path(config.DOWNLOAD_DIR) / video_id
        self.download_dir.mkdir(parents=True, exist_ok=True)
        
        # Track download state
        self.downloaded_segments: Set[int] = set()
        self.failed_segments: Set[int] = set()
        self.session = requests.Session()
        self.session.headers.update({'User-Agent': config.USER_AGENT})
        
        logger.info(f"Initialized downloader for video: {video_id}")
        logger.info(f"Backend URL: {self.backend_url}")
        logger.info(f"Download directory: {self.download_dir}")

    def get_segment_status(self) -> Dict:
        """
        Check which segments are available on the backend
        
        Returns:
            Dict with 'ready' list of available segment numbers
        """
        try:
            url = f"{self.backend_url}/status?v={self.video_id}"
            response = self.session.get(url, timeout=config.REQUEST_TIMEOUT)
            response.raise_for_status()
            
            data = response.json()
            logger.debug(f"Status response: {data}")
            
            return {
                'ready': data.get('ready', []),
                'total': data.get('totalSegments')
            }
            
        except requests.exceptions.RequestException as e:
            logger.error(f"Failed to get segment status: {e}")
            return {'ready': [], 'total': None}
        except json.JSONDecodeError as e:
            logger.error(f"Invalid JSON in status response: {e}")
            return {'ready': [], 'total': None}

    def download_segment(self, segment_num: int, retry_count: int = 0) -> bool:
        """
        Download a specific segment
        
        Args:
            segment_num: Segment number to download
            retry_count: Current retry attempt
            
        Returns:
            True if download successful, False otherwise
        """
        segment_file = self.download_dir / f"segment_{segment_num}.m4a"
        
        # Skip if already exists and is valid
        if segment_file.exists() and segment_file.stat().st_size > 0:
            logger.info(f"Segment {segment_num} already exists, skipping")
            self.downloaded_segments.add(segment_num)
            return True
        
        try:
            url = f"{self.backend_url}/stream?v={self.video_id}&segment={segment_num}"
            logger.info(f"Downloading segment {segment_num} from {url}")
            
            response = self.session.get(url, timeout=config.REQUEST_TIMEOUT, stream=True)
            response.raise_for_status()
            
            # Check content type
            content_type = response.headers.get('content-type', '')
            if not any(ct in content_type for ct in ['audio', 'video', 'octet-stream']):
                logger.warning(f"Unexpected content type for segment {segment_num}: {content_type}")
            
            # Download with progress
            total_size = int(response.headers.get('content-length', 0))
            
            with open(segment_file, 'wb') as f:
                if HAS_TQDM and total_size > 0:
                    with tqdm(total=total_size, unit='B', unit_scale=True, 
                             desc=f"Segment {segment_num}") as pbar:
                        for chunk in response.iter_content(chunk_size=8192):
                            if chunk:
                                f.write(chunk)
                                pbar.update(len(chunk))
                else:
                    for chunk in response.iter_content(chunk_size=8192):
                        if chunk:
                            f.write(chunk)
            
            # Verify download
            if segment_file.stat().st_size == 0:
                segment_file.unlink()
                raise SegmentDownloadError(f"Downloaded segment {segment_num} is empty")
            
            logger.info(f"Successfully downloaded segment {segment_num} ({segment_file.stat().st_size} bytes)")
            self.downloaded_segments.add(segment_num)
            return True
            
        except requests.exceptions.RequestException as e:
            logger.error(f"Network error downloading segment {segment_num}: {e}")
        except Exception as e:
            logger.error(f"Unexpected error downloading segment {segment_num}: {e}")
        
        # Handle retries
        if retry_count < config.MAX_RETRIES:
            wait_time = 2 ** retry_count  # Exponential backoff
            logger.info(f"Retrying segment {segment_num} in {wait_time} seconds (attempt {retry_count + 1}/{config.MAX_RETRIES})")
            time.sleep(wait_time)
            return self.download_segment(segment_num, retry_count + 1)
        else:
            logger.error(f"Failed to download segment {segment_num} after {config.MAX_RETRIES} attempts")
            self.failed_segments.add(segment_num)
            return False

    def download_available_segments(self, max_segments: Optional[int] = None) -> List[int]:
        """
        Download all currently available segments
        
        Args:
            max_segments: Maximum number of segments to download (None for unlimited)
            
        Returns:
            List of successfully downloaded segment numbers
        """
        status = self.get_segment_status()
        available_segments = status['ready']
        
        if not available_segments:
            logger.warning("No segments available for download")
            return []
        
        # Filter out already downloaded segments
        to_download = [seg for seg in available_segments 
                      if seg not in self.downloaded_segments and seg not in self.failed_segments]
        
        if max_segments:
            to_download = to_download[:max_segments]
        
        if not to_download:
            logger.info("All available segments already downloaded")
            return list(self.downloaded_segments)
        
        logger.info(f"Downloading {len(to_download)} segments: {to_download}")
        
        # Download segments concurrently
        successful_downloads = []
        with ThreadPoolExecutor(max_workers=config.MAX_CONCURRENT) as executor:
            future_to_segment = {executor.submit(self.download_segment, seg): seg 
                               for seg in to_download}
            
            for future in as_completed(future_to_segment):
                segment_num = future_to_segment[future]
                try:
                    success = future.result()
                    if success:
                        successful_downloads.append(segment_num)
                except Exception as e:
                    logger.error(f"Thread error for segment {segment_num}: {e}")
        
        return sorted(successful_downloads)

    def monitor_and_download(self, max_wait_time: int = 300) -> List[int]:
        """
        Continuously monitor for new segments and download them
        
        Args:
            max_wait_time: Maximum time to wait in seconds
            
        Returns:
            List of all downloaded segment numbers
        """
        logger.info(f"Starting continuous monitoring (max wait: {max_wait_time}s)")
        start_time = time.time()
        last_segment_count = 0
        no_new_segments_count = 0
        
        while time.time() - start_time < max_wait_time:
            try:
                # Download currently available segments
                downloaded = self.download_available_segments()
                
                if len(downloaded) > last_segment_count:
                    last_segment_count = len(downloaded)
                    no_new_segments_count = 0
                    logger.info(f"Downloaded {len(downloaded)} segments so far")
                else:
                    no_new_segments_count += 1
                
                # Check if we should continue monitoring
                status = self.get_segment_status()
                if status.get('total') and len(self.downloaded_segments) >= status['total']:
                    logger.info("All segments downloaded according to backend")
                    break
                
                # Stop if no new segments for too long
                if no_new_segments_count >= 10:  # 20 seconds without new segments
                    logger.info("No new segments detected for extended period, stopping monitor")
                    break
                
                # Wait before next check
                logger.debug(f"Waiting {config.POLLING_INTERVAL}s before next check...")
                time.sleep(config.POLLING_INTERVAL)
                
            except KeyboardInterrupt:
                logger.info("Monitoring interrupted by user")
                break
            except Exception as e:
                logger.error(f"Error during monitoring: {e}")
                time.sleep(config.POLLING_INTERVAL)
        
        logger.info(f"Monitoring completed. Downloaded {len(self.downloaded_segments)} segments")
        return sorted(self.downloaded_segments)

    def combine_segments(self, output_file: str, use_ffmpeg: bool = True) -> bool:
        """
        Combine downloaded segments into a single audio file
        
        Args:
            output_file: Path to output file
            use_ffmpeg: Whether to use ffmpeg (True) or simple concatenation (False)
            
        Returns:
            True if combination successful
        """
        if not self.downloaded_segments:
            logger.error("No segments downloaded to combine")
            return False
        
        segment_files = []
        for seg_num in sorted(self.downloaded_segments):
            segment_file = self.download_dir / f"segment_{seg_num}.m4a"
            if segment_file.exists():
                segment_files.append(segment_file)
        
        if not segment_files:
            logger.error("No segment files found for combination")
            return False
        
        logger.info(f"Combining {len(segment_files)} segments into {output_file}")
        
        if use_ffmpeg and self._has_ffmpeg():
            return self._combine_with_ffmpeg(segment_files, output_file)
        else:
            return self._combine_simple(segment_files, output_file)

    def _has_ffmpeg(self) -> bool:
        """Check if ffmpeg is available"""
        try:
            import subprocess
            subprocess.run(['ffmpeg', '-version'], capture_output=True, check=True)
            return True
        except (subprocess.CalledProcessError, FileNotFoundError):
            return False

    def _combine_with_ffmpeg(self, segment_files: List[Path], output_file: str) -> bool:
        """Combine segments using ffmpeg"""
        try:
            import subprocess
            
            # Create file list for ffmpeg
            list_file = self.download_dir / "segments_list.txt"
            with open(list_file, 'w') as f:
                for segment_file in segment_files:
                    f.write(f"file '{segment_file.absolute()}'\n")
            
            # Run ffmpeg
            cmd = [
                'ffmpeg', '-y', '-f', 'concat', '-safe', '0',
                '-i', str(list_file), '-c', 'copy', output_file
            ]
            
            logger.info(f"Running ffmpeg: {' '.join(cmd)}")
            result = subprocess.run(cmd, capture_output=True, text=True, check=True)
            
            # Clean up
            list_file.unlink()
            
            logger.info(f"Successfully combined segments with ffmpeg: {output_file}")
            return True
            
        except subprocess.CalledProcessError as e:
            logger.error(f"FFmpeg failed: {e.stderr}")
            return False
        except Exception as e:
            logger.error(f"Error running ffmpeg: {e}")
            return False

    def _combine_simple(self, segment_files: List[Path], output_file: str) -> bool:
        """Simple binary concatenation (may not work for all formats)"""
        try:
            with open(output_file, 'wb') as outfile:
                for segment_file in segment_files:
                    with open(segment_file, 'rb') as infile:
                        outfile.write(infile.read())
            
            logger.info(f"Successfully combined segments (simple): {output_file}")
            return True
            
        except Exception as e:
            logger.error(f"Simple combination failed: {e}")
            return False

    def get_download_stats(self) -> Dict:
        """Get download statistics"""
        return {
            'video_id': self.video_id,
            'downloaded_segments': len(self.downloaded_segments),
            'failed_segments': len(self.failed_segments),
            'downloaded_list': sorted(self.downloaded_segments),
            'failed_list': sorted(self.failed_segments),
            'download_directory': str(self.download_dir)
        }

def main():
    """Main entry point for command line usage"""
    parser = argparse.ArgumentParser(description='Download segmented audio from YouTube video')
    parser.add_argument('video_id', help='YouTube video ID (11 characters)')
    parser.add_argument('--backend-url', default=config.BACKEND_URL, help='Backend server URL')
    parser.add_argument('--download-dir', default=config.DOWNLOAD_DIR, help='Download directory')
    parser.add_argument('--max-wait', type=int, default=300, help='Max wait time in seconds')
    parser.add_argument('--no-monitor', action='store_true', help='Download available segments only, no monitoring')
    parser.add_argument('--combine', help='Combine segments into output file')
    parser.add_argument('--no-ffmpeg', action='store_true', help='Use simple concatenation instead of ffmpeg')
    parser.add_argument('--concurrent', type=int, default=config.MAX_CONCURRENT, help='Max concurrent downloads')
    parser.add_argument('--verbose', '-v', action='store_true', help='Verbose logging')
    
    args = parser.parse_args()
    
    # Configure logging
    if args.verbose:
        logging.getLogger().setLevel(logging.DEBUG)
    
    # Validate video ID
    if not args.video_id or len(args.video_id) != 11:
        logger.error("Invalid video ID. Must be 11 characters (YouTube video ID)")
        sys.exit(1)
    
    # Update configuration
    config.BACKEND_URL = args.backend_url
    config.DOWNLOAD_DIR = args.download_dir
    config.MAX_CONCURRENT = args.concurrent
    
    try:
        # Create downloader
        downloader = SegmentDownloader(args.video_id, args.backend_url)
        
        if args.no_monitor:
            # Download available segments only
            logger.info("Downloading currently available segments...")
            downloaded = downloader.download_available_segments()
        else:
            # Monitor and download continuously
            logger.info("Starting continuous monitoring and download...")
            downloaded = downloader.monitor_and_download(args.max_wait)
        
        # Show statistics
        stats = downloader.get_download_stats()
        logger.info(f"Download complete: {stats}")
        
        # Combine segments if requested
        if args.combine and downloaded:
            success = downloader.combine_segments(args.combine, not args.no_ffmpeg)
            if success:
                logger.info(f"Audio combined successfully: {args.combine}")
            else:
                logger.error("Failed to combine segments")
                sys.exit(1)
        
        if not downloaded:
            logger.warning("No segments were downloaded")
            sys.exit(1)
            
    except KeyboardInterrupt:
        logger.info("Download interrupted by user")
        sys.exit(1)
    except Exception as e:
        logger.error(f"Unexpected error: {e}")
        sys.exit(1)

if __name__ == "__main__":
    main()

# Example usage:
# python segmentDownloader.py dQw4w9WgXcQ --combine output.m4a --verbose
# python segmentDownloader.py dQw4w9WgXcQ --no-monitor --max-concurrent 5